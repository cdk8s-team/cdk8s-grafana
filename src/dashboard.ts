import { Duration } from 'cdk8s';
import { Construct } from 'constructs';
import { GrafanaDashboard } from './imports/grafana-dashboard';

export interface DashboardProps {
  /**
   * Title of the dashboard.
   */
  readonly title: string;

  /**
   * Group dashboards into folders.
   * @default - default folder
   */
  readonly folder?: string;

  /**
   * Specify a mapping from data source variables to data source names.
   * This is only needed if you are importing an existing dashboard's JSON
   * and it specifies variables within an "__inputs" field.
   *
   * @example { DS_PROMETHEUS: "my-prometheus-ds" }
   * @default - no data source variables
   */
  readonly dataSourceVariables?: { [name: string]: string };

  /**
   * Auto-refresh interval.
   * @default - 5 seconds
   */
  readonly refreshRate?: Duration;

  /**
   * Time range for the dashboard, e.g. last 6 hours, last 7 days, etc.
   * @default - 6 hours
   */
  readonly timeRange?: Duration;

  /**
   * Specify plugins required by the dashboard.
   */
  readonly plugins?: GrafanaPlugin[];

  /**
   * Labels to apply to the kubernetes resource.
   *
   * When adding a dashboard to a Grafana instance using `grafana.addDashboard`,
   * labels provided to Grafana will be automatically applied. Otherwise,
   * labels must be added manually.
   *
   * @default - no labels
   */
  readonly labels?: { [name: string]: string };

  /**
   * Namespace to apply to the kubernetes resource.
   *
   * When adding a dashboard to a Grafana instance using `grafana.addDashboard`,
   * the namespace will be automatically inherited.
   *
   * @default - undefined (will be assigned to the 'default' namespace)
   */
  readonly namespace?: string;

  /**
   * All other dashboard customizations.
   * @see https://grafana.com/docs/grafana/latest/dashboards/json-model/
   */
  readonly jsonModel?: { [key: string]: any };
}

/**
 * A Grafana dashboard.
 * @see https://grafana.com/docs/grafana/latest/http_api/dashboard/
 */
export class Dashboard extends Construct {
  private readonly plugins: GrafanaPlugin[];
  constructor(scope: Construct, id: string, props: DashboardProps) {
    super(scope, id);

    this.plugins = [];

    const refreshRate = props.refreshRate ?? Duration.seconds(5);
    const timeRange = props.timeRange ?? Duration.hours(6);
    const dataSources = Object.entries(props.dataSourceVariables ?? {}).map(
      ([variable, name]) => ({ datasourceName: name, inputName: variable }),
    );

    const defaults = {
      title: props.title,
      id: null,
      tags: [],
      style: 'dark',
      timezone: 'browser',
      editable: true,
      hideControls: false,
      graphTooltip: 1,
      panels: [],
      time: {
        from: `now-${timeRange.toSeconds()}s`,
        to: 'now',
      },
      timepicker: {
        time_options: [],
        refresh_intervals: [],
      },
      templating: {
        list: [],
      },
      annotations: {
        list: [],
      },
      refresh: `${refreshRate.toSeconds()}s`,
      schemaVersion: 17,
      version: 0,
      links: [],
    } as any;

    new GrafanaDashboard(this, 'Resource', {
      metadata: {
        labels: props.labels,
        namespace: props.namespace,
      },
      spec: {
        customFolderName: props.folder,
        datasources: dataSources,
        plugins: this.plugins,
        json: JSON.stringify({
          ...defaults,
          ...props.jsonModel,
        }, null, 2),
      },
    });

    if (props.plugins) {
      this.addPlugins(...props.plugins);
    }
  }

  /**
   * Adds one or more plugins.
   */
  public addPlugins(...plugins: GrafanaPlugin[]) {
    for (const plugin of plugins) {
      this.plugins.push(plugin);
    }
  }
}

export interface GrafanaPlugin {
  /**
   * Name of the plugin, e.g. "grafana-piechart-panel"
   */
  readonly name: string;

  /**
   * Version of the plugin, e.g. "1.3.6"
   */
  readonly version: string;
}
