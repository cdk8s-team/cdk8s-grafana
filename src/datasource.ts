import { Construct } from 'constructs';
import { GrafanaDataSource } from './imports/grafana-datasource';

/**
 * Mode for accessing a data source.
 * @see https://grafana.com/docs/grafana/latest/administration/provisioning/#example-data-source-config-file
 */
export enum AccessType {
  /**
   * Access via proxy.
   */
  PROXY = 'proxy',

  /**
   * Access directly (via server or browser in UI).
   */
  DIRECT = 'direct',
}

export interface DataSourceProps {
  /**
   * Name of the data source.
   */
  readonly name: string;

  /**
   * Type of the data source.
   */
  readonly type: string;

  /**
   * Access type of the data source.
   */
  readonly access: AccessType;

  /**
   * Description of the data source.
   * @default - no description
   */
  readonly description?: string;

  /**
   * URL of the data source. Most resources besides the 'testdata' data source
   * type require this field in order to retrieve data.
   *
   * @default - default url for data source type
   */
  readonly url?: string;

  /**
   * Labels to apply to the kubernetes resource.
   *
   * When adding a data source to a Grafana instance using `grafana.addDataSource`,
   * labels provided to Grafana will be automatically applied. Otherwise,
   * labels must be added manually.
   *
   * @default - no labels
   */
  readonly labels?: { [name: string]: string };

  /**
   * Namespace to apply to the kubernetes resource.
   *
   * When adding a data source to a Grafana instance using `grafana.addDataSource`,
   * the namespace will be automatically inherited.
   *
   * @default - undefined (will be assigned to the 'default' namespace)
   */
  readonly namespace?: string;
}

/**
 * A Grafana data source.
 * @see https://grafana.com/docs/grafana/latest/administration/provisioning/#example-data-source-config-file
 */
export class DataSource extends Construct {
  /**
   * Name of the data source.
   */
  public readonly name: string;

  constructor(scope: Construct, id: string, props: DataSourceProps) {
    super(scope, id);

    this.name = props.name;

    new GrafanaDataSource(this, 'Resource', {
      metadata: {
        labels: props.labels,
        namespace: props.namespace,
      },
      spec: {
        name: props.name,
        datasources: [{
          name: props.name,
          type: props.type,
          access: props.access,
          url: props.url,
        }],
      },
    });
  }
}