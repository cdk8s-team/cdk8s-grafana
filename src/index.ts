export * from './dashboard';
export * from './datasource';
export * from './grafana';
