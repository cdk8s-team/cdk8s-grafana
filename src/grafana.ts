import { Construct } from 'constructs';
import { Dashboard, DashboardProps } from './dashboard';
import { DataSource, DataSourceProps } from './datasource';
import { Grafana as RawGrafana } from './imports/grafana';
import { LabelSelector } from './imports/k8s';

export interface GrafanaProps {
  /**
   * Specify a custom image for Grafana.
   * @default "public.ecr.aws/ubuntu/grafana:latest"
   */
  readonly image?: string;

  /**
   * Create an ingress to provide external access to the Grafana cluster.
   * @default true
   */
  readonly ingress?: boolean;

  /**
   * Type of service to be created (NodePort, ClusterIP or LoadBalancer)
   * @default ClusterIP
   */
  readonly serviceType?: string;

  /**
   * Default admin username.
   * @default "root"
   */
  readonly adminUser?: string;

  /**
   * Default admin password.
   * @default "secret"
   */
  readonly adminPassword?: string;

  /**
   * Require login in order to view or manage dashboards.
   * @default false
   */
  readonly requireLogin?: boolean;

  /**
   * Default data source - equivalent to calling `grafana.addDataSource`.
   * @default - no data source added
   */
  readonly defaultDataSource?: DataSourceProps;

  /**
   * Labels to apply to all Grafana resources.
   * @default - { app: "grafana" }
   */
  readonly labels?: { [name: string]: string };

  /**
   * Namespace to apply to all Grafana resources. The Grafana Operator must be
   * installed in this namespace for resources to be recognized.
   *
   * @default - undefined (will be assigned to the 'default' namespace)
   */
  readonly namespace?: string;
}

/**
 * A Grafana instance.
 */
export class Grafana extends Construct {
  private readonly dataSources: DataSource[];
  private readonly dashboards: Dashboard[];
  private readonly namespace: string | undefined;
  private readonly labels: { [name: string]: string };

  constructor(scope: Construct, id: string, props: GrafanaProps = {}) {
    super(scope, id);

    this.labels = props.labels ?? { app: 'grafana' };
    this.dataSources = [];
    this.dashboards = [];
    this.namespace = props.namespace;

    const baseImage = props.image ?? 'public.ecr.aws/ubuntu/grafana:latest';
    const ingress = props.ingress ?? true;
    const serviceType = props.serviceType ?? 'ClusterIP';
    const adminUser = props.adminUser ?? 'root';
    const adminPassword = props.adminPassword ?? 'secret';
    const requireLogin = props.requireLogin ?? false;
    const dashboardLabelSelectors: LabelSelector[] = [{ matchLabels: this.labels ?? { app: 'grafana' } }];

    new RawGrafana(this, 'Resource', {
      metadata: {
        labels: this.labels,
        namespace: this.namespace,
      },
      spec: {
        baseImage: baseImage,
        ingress: {
          enabled: ingress,
        },
        service: {
          type: serviceType,
        },
        client: {
          // without this, dashboards may not be automatically discovered
          // https://github.com/grafana-operator/grafana-operator/blob/master/documentation/deploy_grafana.md#configuring-grafana-api-access
          preferService: true,
        },
        config: {
          'log': {
            mode: 'console',
            level: 'info',
          },
          'auth.anonymous': {
            enabled: !requireLogin,
          },
          'security': {
            admin_user: adminUser,
            admin_password: adminPassword,
          },
        },
        deployment: {
          labels: this.labels,
        },
        dashboardLabelSelector: dashboardLabelSelectors,
      },
    });

    if (props.defaultDataSource) {
      this.addDataSource('default-datasource', props.defaultDataSource);
    }
  }

  /**
   * Adds a data source. By default, labels are automatically added so that
   * the data source is detected by Grafana.
   */
  public addDataSource(id: string, props: DataSourceProps): DataSource {
    const labels = {
      ...this.labels,
      ...props.labels,
    };
    const datasource = new DataSource(this, id, {
      labels,
      namespace: this.namespace,
      ...props,
    });
    this.dataSources.push(datasource);
    return datasource;
  }

  /**
   * Creates a dashboard associated with a particular data source. By default,
   * labels are automatically added so that the data source is detected by
   * Grafana.
   */
  public addDashboard(id: string, props: DashboardProps): Dashboard {
    const labels = {
      ...this.labels,
      ...props.labels,
    };
    const dashboard = new Dashboard(this, id, {
      labels,
      namespace: this.namespace,
      ...props,
    });
    this.dashboards.push(dashboard);
    return dashboard;
  }
}
