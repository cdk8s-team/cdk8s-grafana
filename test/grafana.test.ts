import { Chart, Duration, Testing } from 'cdk8s';
import { AccessType, Grafana } from '../src';

const PLUGINS = [
  {
    name: 'grafana-piechart-panel',
    version: '1.3.6',
  },
  {
    name: 'grafana-clock-panel',
    version: '1.0.2',
  },
];

describe('a grafana instance', () => {
  test('defaults', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new Grafana(chart, 'my-grafana');

    // THEN
    expect(Testing.synth(chart)).toMatchSnapshot();
  });

  test('Nodeport and no ingress', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new Grafana(chart, 'my-grafana', {
      ingress: false,
      serviceType: 'NodePort',
    });

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest).toMatchSnapshot();
    expect(manifest[0].spec.ingress.enabled).toEqual(false);
    expect(manifest[0].spec.service.type).toEqual('NodePort');
  });

  test('with customized auth', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new Grafana(chart, 'my-grafana', {
      adminUser: 'admin',
      adminPassword: 'this-is-a-bad-password',
      requireLogin: true,
    });

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest).toMatchSnapshot();
    expect(manifest[0].spec.config['auth.anonymous'].enabled).toEqual(false);
    expect(manifest[0].spec.config.security.admin_user).toEqual('admin');
    expect(manifest[0].spec.config.security.admin_password).toEqual('this-is-a-bad-password');
  });

  test('with data source and customized dashboard', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    const grafana = new Grafana(chart, 'my-grafana', {
      defaultDataSource: {
        name: 'Prometheus',
        type: 'prometheus',
        access: AccessType.PROXY,
        url: 'http://prometheus-service:9090',
      },
    });
    grafana.addDashboard('my-dashboard', {
      title: 'My Dashboard',
      folder: 'special-dashboards',
      refreshRate: Duration.seconds(10),
      timeRange: Duration.hours(6),
      jsonModel: {
        hideControls: true,
        panels: [
          {
            type: 'text',
            title: 'Panel Title',
            gridPos: {
              x: 0,
              y: 0,
              w: 12,
              h: 9,
            },
            mode: 'markdown',
            content: '# title',
          },
        ],
      },
    });

    // THEN
    expect(Testing.synth(chart)).toMatchSnapshot();
  });

  test('with a dashboard expecting a data source variable', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    const grafana = new Grafana(chart, 'my-grafana');
    const prometheus = grafana.addDataSource('prometheus', {
      name: 'Prometheus',
      type: 'prometheus',
      access: AccessType.PROXY,
      url: 'http://prometheus-service:9090',
    });
    grafana.addDashboard('my-dashboard', {
      title: 'My Dashboard',
      folder: 'special-dashboards',
      refreshRate: Duration.seconds(10),
      timeRange: Duration.hours(6),
      jsonModel: {
        __inputs: [
          {
            name: 'DS_PROMETHEUS',
            label: 'Prometheus',
            description: '',
            type: 'datasource',
            pluginId: 'prometheus',
            pluginName: 'Prometheus',
          },
        ],
      },
      dataSourceVariables: {
        DS_PROMETHEUS: prometheus.name,
      },
    });

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest).toMatchSnapshot();
    expect(manifest[2].spec.datasources.length).toEqual(1);
  });

  test('labels are inherited by dashboards and data sources', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    const grafana = new Grafana(chart, 'my-grafana', {
      labels: { stage: 'prod' },
    });
    grafana.addDashboard('my-dashboard', {
      title: 'My Dashboard',
    });
    grafana.addDataSource('prometheus', {
      name: 'Prometheus',
      type: 'prometheus',
      access: AccessType.PROXY,
      url: 'http://prometheus-service:9090',
    });

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest).toMatchSnapshot();
    expect(manifest[0].metadata.labels.stage).toEqual('prod'); // grafana instance
    expect(manifest[1].metadata.labels.stage).toEqual('prod'); // dashboard instance
    expect(manifest[2].metadata.labels.stage).toEqual('prod'); // data source instance
  });

  test('namespaces are inherited by dashboards and data sources', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    const grafana = new Grafana(chart, 'my-grafana', {
      namespace: 'my-namespace',
    });
    grafana.addDashboard('my-dashboard', {
      title: 'My Dashboard',
    });
    grafana.addDataSource('prometheus', {
      name: 'Prometheus',
      type: 'prometheus',
      access: AccessType.PROXY,
      url: 'http://prometheus-service:9090',
    });

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0].metadata.namespace).toEqual('my-namespace'); // grafana instance
    expect(manifest[1].metadata.namespace).toEqual('my-namespace'); // dashboard instance
    expect(manifest[2].metadata.namespace).toEqual('my-namespace'); // data source instance
  });

  test('adding plugins via constructor', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    const grafana = new Grafana(chart, 'my-grafana');
    grafana.addDashboard('my-dashboard', {
      title: 'My Dashboard',
      plugins: PLUGINS,
    });

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[1].spec.plugins).toMatchObject(PLUGINS);
  });

  test('adding plugins via method', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    const grafana = new Grafana(chart, 'my-grafana');
    const dashboard = grafana.addDashboard('my-dashboard', { title: 'My Dashboard' });
    dashboard.addPlugins(...PLUGINS);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[1].spec.plugins).toMatchObject(PLUGINS);
  });
});
