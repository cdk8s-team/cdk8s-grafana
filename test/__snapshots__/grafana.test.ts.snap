// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a grafana instance defaults 1`] = `
Array [
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "Grafana",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-c8153986",
    },
    "spec": Object {
      "baseImage": "public.ecr.aws/ubuntu/grafana:latest",
      "client": Object {
        "preferService": true,
      },
      "config": Object {
        "auth.anonymous": Object {
          "enabled": true,
        },
        "log": Object {
          "level": "info",
          "mode": "console",
        },
        "security": Object {
          "admin_password": "secret",
          "admin_user": "root",
        },
      },
      "dashboardLabelSelector": Array [
        Object {
          "matchLabels": Object {
            "app": "grafana",
          },
        },
      ],
      "deployment": Object {
        "labels": Object {
          "app": "grafana",
        },
      },
      "ingress": Object {
        "enabled": true,
      },
    },
  },
]
`;

exports[`a grafana instance labels are inherited by dashboards and data sources 1`] = `
Array [
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "Grafana",
    "metadata": Object {
      "labels": Object {
        "stage": "prod",
      },
      "name": "test-my-grafana-c8153986",
    },
    "spec": Object {
      "baseImage": "public.ecr.aws/ubuntu/grafana:latest",
      "client": Object {
        "preferService": true,
      },
      "config": Object {
        "auth.anonymous": Object {
          "enabled": true,
        },
        "log": Object {
          "level": "info",
          "mode": "console",
        },
        "security": Object {
          "admin_password": "secret",
          "admin_user": "root",
        },
      },
      "dashboardLabelSelector": Array [
        Object {
          "matchLabels": Object {
            "stage": "prod",
          },
        },
      ],
      "deployment": Object {
        "labels": Object {
          "stage": "prod",
        },
      },
      "ingress": Object {
        "enabled": true,
      },
    },
  },
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "GrafanaDashboard",
    "metadata": Object {
      "labels": Object {
        "stage": "prod",
      },
      "name": "test-my-grafana-my-dashboard-c83fc47b",
    },
    "spec": Object {
      "datasources": Array [],
      "json": "{
  \\"title\\": \\"My Dashboard\\",
  \\"id\\": null,
  \\"tags\\": [],
  \\"style\\": \\"dark\\",
  \\"timezone\\": \\"browser\\",
  \\"editable\\": true,
  \\"hideControls\\": false,
  \\"graphTooltip\\": 1,
  \\"panels\\": [],
  \\"time\\": {
    \\"from\\": \\"now-21600s\\",
    \\"to\\": \\"now\\"
  },
  \\"timepicker\\": {
    \\"time_options\\": [],
    \\"refresh_intervals\\": []
  },
  \\"templating\\": {
    \\"list\\": []
  },
  \\"annotations\\": {
    \\"list\\": []
  },
  \\"refresh\\": \\"5s\\",
  \\"schemaVersion\\": 17,
  \\"version\\": 0,
  \\"links\\": []
}",
      "name": "my-dashboard",
      "plugins": Array [],
    },
  },
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "GrafanaDataSource",
    "metadata": Object {
      "labels": Object {
        "stage": "prod",
      },
      "name": "test-my-grafana-prometheus-c8a66b3d",
    },
    "spec": Object {
      "datasources": Array [
        Object {
          "access": "proxy",
          "name": "Prometheus",
          "type": "prometheus",
          "url": "http://prometheus-service:9090",
        },
      ],
      "name": "Prometheus",
    },
  },
]
`;

exports[`a grafana instance with a dashboard expecting a data source variable 1`] = `
Array [
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "Grafana",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-c8153986",
    },
    "spec": Object {
      "baseImage": "public.ecr.aws/ubuntu/grafana:latest",
      "client": Object {
        "preferService": true,
      },
      "config": Object {
        "auth.anonymous": Object {
          "enabled": true,
        },
        "log": Object {
          "level": "info",
          "mode": "console",
        },
        "security": Object {
          "admin_password": "secret",
          "admin_user": "root",
        },
      },
      "dashboardLabelSelector": Array [
        Object {
          "matchLabels": Object {
            "app": "grafana",
          },
        },
      ],
      "deployment": Object {
        "labels": Object {
          "app": "grafana",
        },
      },
      "ingress": Object {
        "enabled": true,
      },
    },
  },
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "GrafanaDataSource",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-prometheus-c8a66b3d",
    },
    "spec": Object {
      "datasources": Array [
        Object {
          "access": "proxy",
          "name": "Prometheus",
          "type": "prometheus",
          "url": "http://prometheus-service:9090",
        },
      ],
      "name": "Prometheus",
    },
  },
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "GrafanaDashboard",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-my-dashboard-c83fc47b",
    },
    "spec": Object {
      "customFolderName": "special-dashboards",
      "datasources": Array [
        Object {
          "datasourceName": "Prometheus",
          "inputName": "DS_PROMETHEUS",
        },
      ],
      "json": "{
  \\"title\\": \\"My Dashboard\\",
  \\"id\\": null,
  \\"tags\\": [],
  \\"style\\": \\"dark\\",
  \\"timezone\\": \\"browser\\",
  \\"editable\\": true,
  \\"hideControls\\": false,
  \\"graphTooltip\\": 1,
  \\"panels\\": [],
  \\"time\\": {
    \\"from\\": \\"now-21600s\\",
    \\"to\\": \\"now\\"
  },
  \\"timepicker\\": {
    \\"time_options\\": [],
    \\"refresh_intervals\\": []
  },
  \\"templating\\": {
    \\"list\\": []
  },
  \\"annotations\\": {
    \\"list\\": []
  },
  \\"refresh\\": \\"10s\\",
  \\"schemaVersion\\": 17,
  \\"version\\": 0,
  \\"links\\": [],
  \\"__inputs\\": [
    {
      \\"name\\": \\"DS_PROMETHEUS\\",
      \\"label\\": \\"Prometheus\\",
      \\"description\\": \\"\\",
      \\"type\\": \\"datasource\\",
      \\"pluginId\\": \\"prometheus\\",
      \\"pluginName\\": \\"Prometheus\\"
    }
  ]
}",
      "name": "my-dashboard",
      "plugins": Array [],
    },
  },
]
`;

exports[`a grafana instance with customized auth 1`] = `
Array [
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "Grafana",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-c8153986",
    },
    "spec": Object {
      "baseImage": "public.ecr.aws/ubuntu/grafana:latest",
      "client": Object {
        "preferService": true,
      },
      "config": Object {
        "auth.anonymous": Object {
          "enabled": false,
        },
        "log": Object {
          "level": "info",
          "mode": "console",
        },
        "security": Object {
          "admin_password": "this-is-a-bad-password",
          "admin_user": "admin",
        },
      },
      "dashboardLabelSelector": Array [
        Object {
          "matchLabels": Object {
            "app": "grafana",
          },
        },
      ],
      "deployment": Object {
        "labels": Object {
          "app": "grafana",
        },
      },
      "ingress": Object {
        "enabled": true,
      },
    },
  },
]
`;

exports[`a grafana instance with data source and customized dashboard 1`] = `
Array [
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "Grafana",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-c8153986",
    },
    "spec": Object {
      "baseImage": "public.ecr.aws/ubuntu/grafana:latest",
      "client": Object {
        "preferService": true,
      },
      "config": Object {
        "auth.anonymous": Object {
          "enabled": true,
        },
        "log": Object {
          "level": "info",
          "mode": "console",
        },
        "security": Object {
          "admin_password": "secret",
          "admin_user": "root",
        },
      },
      "dashboardLabelSelector": Array [
        Object {
          "matchLabels": Object {
            "app": "grafana",
          },
        },
      ],
      "deployment": Object {
        "labels": Object {
          "app": "grafana",
        },
      },
      "ingress": Object {
        "enabled": true,
      },
    },
  },
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "GrafanaDataSource",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-datasource-c8bd3b25",
    },
    "spec": Object {
      "datasources": Array [
        Object {
          "access": "proxy",
          "name": "Prometheus",
          "type": "prometheus",
          "url": "http://prometheus-service:9090",
        },
      ],
      "name": "Prometheus",
    },
  },
  Object {
    "apiVersion": "integreatly.org/v1alpha1",
    "kind": "GrafanaDashboard",
    "metadata": Object {
      "labels": Object {
        "app": "grafana",
      },
      "name": "test-my-grafana-my-dashboard-c83fc47b",
    },
    "spec": Object {
      "customFolderName": "special-dashboards",
      "datasources": Array [],
      "json": "{
  \\"title\\": \\"My Dashboard\\",
  \\"id\\": null,
  \\"tags\\": [],
  \\"style\\": \\"dark\\",
  \\"timezone\\": \\"browser\\",
  \\"editable\\": true,
  \\"hideControls\\": true,
  \\"graphTooltip\\": 1,
  \\"panels\\": [
    {
      \\"type\\": \\"text\\",
      \\"title\\": \\"Panel Title\\",
      \\"gridPos\\": {
        \\"x\\": 0,
        \\"y\\": 0,
        \\"w\\": 12,
        \\"h\\": 9
      },
      \\"mode\\": \\"markdown\\",
      \\"content\\": \\"# title\\"
    }
  ],
  \\"time\\": {
    \\"from\\": \\"now-21600s\\",
    \\"to\\": \\"now\\"
  },
  \\"timepicker\\": {
    \\"time_options\\": [],
    \\"refresh_intervals\\": []
  },
  \\"templating\\": {
    \\"list\\": []
  },
  \\"annotations\\": {
    \\"list\\": []
  },
  \\"refresh\\": \\"10s\\",
  \\"schemaVersion\\": 17,
  \\"version\\": 0,
  \\"links\\": []
}",
      "name": "my-dashboard",
      "plugins": Array [],
    },
  },
]
`;
